src/iterators/intersection_iterator.c

```diff
@@ -249,51 +249,51 @@ static void II_SetEstimation(IntersectionIterator *it) {
 /**
  * Reduce the intersection iterator by applying these rules:
  * 1. If any of the iterators is an empty iterator, return the empty iterator and update the number of children
- * 2. Remove all wildcard iterators since they would not contribute to the intersection
+ * 2. Remove all wildcard iterators since they would not contribute to the intersection (Return one of them if all are wildcards)
  * 3. If there is only one left child iterator, return it
  * 4. Otherwise, return NULL and let the caller create the intersection iterator
 */
-static QueryIterator *IntersectionIteratorReducer(QueryIterator **its, size_t *num, int max_slop, bool in_order, double weight) {
+static QueryIterator *IntersectionIteratorReducer(QueryIterator **its, size_t *num) {
   QueryIterator *ret = NULL;
 
   // Remove all wildcard iterators from the array
   size_t current_size = *num;
   size_t write_idx = 0;
   size_t empty_iterator_idx = 0;
-  bool has_empty_iterator = false;
+  bool all_wildcards = true;
   for (size_t read_idx = 0; read_idx < current_size; read_idx++) {
     if (IsWildcardIterator(its[read_idx])) {
-      its[read_idx]->Free(its[read_idx]);
+      if (!all_wildcards || all_wildcards && read_idx != current_size - 1) {
+        // remove all the wildcards in case there are other non-wildcard iterators
+        // avoid removing it in case it's the last one and all are wildcards
+        its[read_idx]->Free(its[read_idx]);
+      }
     } else {
+      all_wildcards = false;
       its[write_idx++] = its[read_idx];
     }
   }
   *num = write_idx;
 
   for (size_t ii = 0; ii < write_idx; ++ii) {
     if (!its[ii] || its[ii]->type == EMPTY_ITERATOR) {
-      has_empty_iterator = true;
       empty_iterator_idx = ii;
```

Comment:
If we "take" the iterator here (set `its[ii] = NULL`), we don't need to track this index, and we can simplify the rest of the logic